File: civitasRUBY/lib/mazo_sorpresas.rb
#encoding: utf-8

module Civitas 
  class MazoSorpresas
    
    attr_accessor :usadas, :ultimaSorpresa, :sorpresas
    
    def initialize(_debug=false)
      @sorpresas = Array.new
      @barajada = false
      @usadas = 0
      @cartasEspeciales = Array.new
      @ultimaSorpresa = nil
      @debug = _debug
    end
    
    public
    def alMazo (nueva)
      if (!@barajada)
        @sorpresas.push(nueva)
      end
    end
    
    #Devuelve la siguiente sorpresa, barajando si no está barajado
    def siguiente
      if (!@barajada || @usadas==@sorpresas.size) && !@debug
        @sorpresas.shuffle #Baraja
        @usadas   = 0
        @barajada = true
      end
      
      @usadas += 1
      
      @ultimaSorpresa = @sorpresas[0] # [1, 2, 3]
      @sorpresas << @sorpresas[0]     # [1, 2, 3, 1]
      @sorpresas.delete_at(0)         # [2, 3, 1]
      
      return @ultimaSorpresa      
    end
    
    #Inhabilita carta sorpresa de mazo, convirtiéndola en especial
    def inhabilitarCartaEspecial(sorpresa)
      for carta in @sorpresas
        if @sopresas==sorpresa
          @cartasEspeciales << carta
          @sorpresas.delete(carta)
          
          Diario.instance.ocurreEvento("Carta especial inhabilitada")
        end
      end
    end
    
    #Habilita carta sorpresa de mazo, convirtiéndola en carta común
    def habilitarCartaEspecial(sorpresa)
      for carta in @cartasEspeciales
        if @cartasEspeciales==sorpresa
          @sorpresas << sorpresa
          @cartasEspeciales.delete(carta)
          
          Diario.instance.ocurreEvento("Carta especial habilitada")
        end
      end
    end
    
  end
end

File: civitasJAVA/civitas/src/civitas/MazoSorpresas.java
package civitas;

import java.util.ArrayList;
import java.util.Collections;


public class MazoSorpresas {
    private Boolean barajada;
    private Boolean debug;
    private Sorpresa ultimaSorpresa;
    private int usadas;

    private ArrayList<Sorpresa> sorpresas;
    private ArrayList<Sorpresa> cartasEspeciales;


    //Constructor debug
    public MazoSorpresas(Boolean debug) {
        this.debug = debug;
        init();
        if (debug == true) {
            Diario.getInstance().ocurreEvento("MazoSorpresa debug true");
        }
    }

    //Constructor sin parámetros
    public MazoSorpresas() {
        init();
        debug = false;
    }

    //Método de inicialización común a todos los constructores
    private void init() {
        sorpresas = new ArrayList<>();
        cartasEspeciales = new ArrayList<>();
        barajada = false;
        usadas = 0;
    }
    
    //Añade la sorpresa s al mazo si no ha sido barajado
    public void alMazo(Sorpresa sorpresa) {
        if (barajada == false) {
            sorpresas.add(sorpresa);
        }
    }

    //Mecánica para activación de carta especial
    public void habilitarCartaEspecial(Sorpresa sorpresa) {
        Boolean encontrada = false;
        for (int i = 0; i < cartasEspeciales.size() && !encontrada; i++) {
            if (cartasEspeciales.get(i) == sorpresa) {
                encontrada = true;
                cartasEspeciales.remove(i);
                sorpresas.add(sorpresa);
                Diario.getInstance().ocurreEvento("Carta especial habilitada");
            }
        }
    }

    //Inhabilita carta sorpresa de mazo, convirtiéndola en especial
    public void inhabilitarCartaEspecial(Sorpresa sorpresa) {
        Boolean encontrada = false;
        for (int i = 0; i < sorpresas.size() && !encontrada; i++) {
            if (sorpresas.get(i) == sorpresa) {
                encontrada = true;
                sorpresas.remove(i);
                cartasEspeciales.add(sorpresa);
                Diario.getInstance().ocurreEvento("Carta especial inhabilitada");
            }
        }
    }

    //Devuelve la siguiente sorpresa, barajando si no está barajado
    public Sorpresa siguiente() {
        if (barajada == false || usadas == sorpresas.size()) {
            if (debug == false) {
                Collections.shuffle(sorpresas);
                usadas = 0;
                barajada = true;
            }

        }

        usadas++;
        ultimaSorpresa = sorpresas.get(0);
        sorpresas.remove(0);
        sorpresas.add(ultimaSorpresa);

        return ultimaSorpresa;
    }

}
